### PR TITLE
Feat: Create tx manager and run functionality

### PR DESCRIPTION
**Pull request by Shailu-s**

### CODE REVIEW

File: sequencer/coordinator/txmanager.go [hunk 1]
```diff
@@ -132,4 +136,543 @@ func (t *TxManager) AddBatch(ctx context.Context, batchInfo *BatchInfo) {
...
+	receipt := batchInfo.Receipt
+	if receipt != nil {
+		// taking last eth tx nonce as it probably have the highest nonce
+		lastEthTx := batchInfo.EthTxs[len(batchInfo.EthTxs)-1]
+		if lastEthTx.Nonce()+1 > t.accNonce {
+			t.accNonce = lastEthTx.Nonce() + 1
+		}
+		if receipt.Status == types.ReceiptStatusFailed {
+			batchInfo.Debug.Status = StatusFailed
+			_, err := t.ethClient.EthCall(ctx, lastEthTx, receipt.BlockNumber)
```
Comment (Intizar-T) [inline sequencer/coordinator/txmanager.go:224]: the definition of `EthCall` says 
```go
// EthCall runs the transaction as a call (without paying) in the local node at
// blockNum
```
but we don't have a local node, right? Could you pls take a closer look into this? 🙏

**Comment by Intizar-T** [comment]
@Shailu-s @kaiden-araki any idea why the TxManager keeps manual track of the nonce? That's a dangerous operation as skipping a single nonce will not allow any further txs to be accepted. In my prev company we used to have a dedicated manual nonce manager package and it used to be big deal to build and operate it as it can really mess things up.

File: sequencer/coordinator/txmanager.go [hunk 2]
```diff
@@ -132,4 +136,543 @@ func (t *TxManager) AddBatch(ctx context.Context, batchInfo *BatchInfo) {
...
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("TxManager done")
+			return
+		case statsVars := <-t.statsVarsCh:
+			t.stats = statsVars.Stats
+			t.syncSCVars(statsVars.Vars)
+		case pipelineNum := <-t.discardPipelineCh:
+			t.minPipelineNum = pipelineNum + 1
```
Comment (Intizar-T) [inline sequencer/coordinator/txmanager.go:558]: Given that we'll have a single prover server, we might use only one pipeline, but I'm not sure. Need to look more into this. @Shailu-s could you pls either look into it as part of this ticket/PR or create a separate notion ticket in the backlog for this? 🙏

File: sequencer/coordinator/txmanager.go [hunk 3]
```diff
@@ -132,4 +136,543 @@ func (t *TxManager) AddBatch(ctx context.Context, batchInfo *BatchInfo) {
...
+			if err := t.removeBadBatchInfos(ctx); ctx.Err() != nil {
+				continue
+			} else if err != nil {
+				log.Errorw("TxManager: removeBadBatchInfos", "err", err)
+				continue
+			}
+		case batchInfo := <-t.batchCh:
+			if batchInfo.BatchNum > t.lastSuccessBatch+1 && t.lastSuccessBatch != 0 {
+				go func(t *TxManager, batchInfo *BatchInfo) {
+					const waitIntervalBeforeWritingBatchToChannel = 500
```
Comment (Intizar-T) [inline sequencer/coordinator/txmanager.go]: how about moving this const to a more visible place? either to the top of the file or cfg

File: sequencer/coordinator/txmanager.go [hunk 4]
```diff
@@ -132,4 +136,543 @@ func (t *TxManager) AddBatch(ctx context.Context, batchInfo *BatchInfo) {
...
+					log.Errorw("TxManager: removeBadBatchInfos", "err", err)
+					continue
+				}
+				t.coord.SendMsg(ctx, MsgStopPipeline{
+					Reason: fmt.Sprintf("forgeBatch reject: %v", err),
+				})
+				continue
+			}
+			now := time.Now()
+			if !t.cfg.EthNoReuseNonce && confirm == nil &&
```
Comment (Intizar-T) [inline sequencer/coordinator/txmanager.go:644]: do we go into this condition? if not, lets comment it out for now

File: sequencer/coordinator/txmanager.go [hunk 5]
```diff
@@ -132,4 +136,543 @@ func (t *TxManager) AddBatch(ctx context.Context, batchInfo *BatchInfo) {
...
+	auth.Context = ctx
+	for attempt := 0; attempt < t.cfg.EthClientAttempts; attempt++ {
+		maxGasPrice := big.NewInt(t.cfg.MaxGasPrice)
+		maxGasPrice.Mul(maxGasPrice, big.NewInt(params.GWei))
+		if auth.GasPrice.Cmp(maxGasPrice) > 0 {
+			return fmt.Errorf("calculated gasPrice (%v) > maxGasPrice (%v)",
+				auth.GasPrice, t.cfg.MaxGasPrice)
+		}
+		// RollupForgeBatch() calls ethclient.SendTransaction()
+		// ethTx, err = t.ethClient.RollupForgeBatch(batchInfo.ForgeBatchArgs, auth)
```
Comment (Intizar-T) [inline sequencer/coordinator/txmanager.go:452]: i wonder why this is commented out 🤔

File: sequencer/coordinator/txmanager.go [hunk 6]
```diff
@@ -132,4 +136,543 @@ func (t *TxManager) AddBatch(ctx context.Context, batchInfo *BatchInfo) {
...
+		// We check the errors via strings because we match the
+		// definition of the error from geth, with the string returned
+		// via RPC obtained by the client.
+		if err == nil {
+			break
+		} else if strings.Contains(err.Error(), core.ErrNonceTooLow.Error()) {
+			log.Warnw("TxManager ethClient.RollupForgeBatch incrementing nonce",
+				"err", err, "nonce", auth.Nonce, "batchNum", batchInfo.BatchNum)
+			auth.Nonce.Add(auth.Nonce, big.NewInt(1))
```
Comment (Intizar-T) [inline sequencer/coordinator/txmanager.go:461]: ugh i really don't like this kind of nonce management style but I guess as long as it works. I'm surprised hermez didn't implement a nonce manager package 🤔

**Comment by Shailu-s** [comment]
> @Shailu-s @kaiden-araki any idea why the TxManager keeps manual track of the nonce? That's a dangerous operation as skipping a single nonce will not allow any further txs to be accepted. In my prev company we used to have a dedicated manual nonce manager package and it used to be big deal to build and operate it as it can really mess things up.

I think nonce is coming here from block info which infact fetches it from curernt block number. It its not breaking in hermez then I think it will work here too.

File: sequencer/coordinator/txmanager.go [hunk 7]
```diff
@@ -132,4 +136,543 @@ func (t *TxManager) AddBatch(ctx context.Context, batchInfo *BatchInfo) {
...
+	// doesn't result in the same value, making the transaction to be equal
+	// than before.
+	if r.Cmp(big.NewInt(0)) == 0 {
+		r = big.NewInt(1)
+	}
+	return r.Add(v, r)
+}
+
+// NewAuth generates a new auth object for an ethereum transaction
+func (t *TxManager) NewAuth(ctx context.Context, batchInfo *BatchInfo) (*bind.TransactOpts, error) {
```
Comment (kaiden-araki) [inline sequencer/coordinator/txmanager.go:351]: @Shailu-s Why do we need NewAuth function? Could you elaborate the reason?

Comment (Shailu-s) [reply sequencer/coordinator/txmanager.go:351]: New auth is a function to get basic requirements for a transaction like gasfees, prices and other details. Actually TransactOpts is compulsory for each txn of ethereum but sometime web3 libraries and wallet like metamask send defualt values. Here this functions will help in more precise calculation.

**Review by kaiden-araki — DISMISSED** [review]

File: sequencer/coordinator/txmanager.go (hunk 6 above)
Comment (Shailu-s) [reply sequencer/coordinator/txmanager.go:461]: We will implement it later if needed.

File: sequencer/coordinator/txmanager.go (hunk 1 above)
Comment (Shailu-s) [reply sequencer/coordinator/txmanager.go:224]: Local Node Usage: While eth_call can be used with a local Ethereum node, it is not limited to local nodes only. You can also use eth_call with any Ethereum node that you can connect to via RPC, whether it's a local node you run yourself or a remote node provided by services like Infura or Alchemy. The key requirement is that the node must support the JSON-RPC interface, which includes the eth_call method. 

This ethCall function is basically just to execute message on blockchain without running the transaction, kind of simulating the transaction.

File: sequencer/coordinator/txmanager.go (hunk 3 above)
Comment (Shailu-s) [reply sequencer/coordinator/txmanager.go]: Done

File: sequencer/coordinator/txmanager.go (hunk 2 above)
Comment (Shailu-s) [reply sequencer/coordinator/txmanager.go:558]: This is basically for times when a pipeline is discarded, then increasing minimum pilelines for one. It seems correct for now, but I will create the ticket and pick it up for more look into use of pipelines and optmization.

File: sequencer/coordinator/txmanager.go (hunk 4 above)
Comment (Shailu-s) [reply sequencer/coordinator/txmanager.go:644]: I think its an important case which is check if forging failed due to reasons or we can reuse nonce if tx is pending. I think we should keep it.

File: sequencer/coordinator/txmanager.go (hunk 5 above)
Comment (Shailu-s) [reply sequencer/coordinator/txmanager.go:452]: Actually this function is currently commented out in rollup.go file so cant be used here. I will cover this under forge batch ticket.

Comment (Intizar-T) [reply sequencer/coordinator/txmanager.go:452]: sounds good, just a minor suggestion to leave a 
```go
// TODO: implement
```
comment to not accidentally miss this

Comment (Shailu-s) [reply sequencer/coordinator/txmanager.go:452]: Done

**Review by Intizar-T — APPROVED** [review]
lgtm! 🔥

**Pull request merged**